### ISSUE TITLE
Don't report for-of variable as unused

### ISSUE THREAD

**Issue by mykhalov** [eslint/eslint#14868]
**What rule do you want to change?**
no-unused-vars

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behaviour, possibly configurable

**Please provide some example code that this change will affect:**

```js
for (const _ of iteratorWithSideEffects()) {
  if (condition()) break
}
```

**What does the rule currently do for this code?**
Reports an error: `'_' is assigned a value but never used.`

**What will the rule do after it's changed?**
Don't report an error

**Are you willing to submit a pull request to implement this change?**
Maybe

---

Not 'using' the variable of for-of loop isn't necessary an error, e.g. if iterator has side effects and yielded values don't matter.

**Comment by nzakas** [comment]
Thanks for the suggestion. We don't make changes based on variable naming conventions because they are very personal. Your best bet in this case is just to use a disable comment where you use a pattern like this.

**Comment by nzakas** [comment]
Oh, you can also use the [`varsIgnorePattern`](https://eslint.org/docs/rules/no-unused-vars#varsignorepattern) option to specify to ignore `_`.

**Comment by mykhalov** [comment]
It's not about naming convention, though. To put it in perspective: if one wants to discard a returned value of a function, it's possible to skip the assignment of the result:

```
const _ = functionWithSideEffect()
```

can be just

```
functionWithSideEffect()
```

One can't, however, exhaust an iterator with for-of loop without assigning the result to something:

```
for (const _ of iteratorWithSideEffects()) {}
```

the following won't work

```
for (iteratorWithSideEffects()) {}
```

It would be nice to have an option to consider the variable 'used' here, regardless on naming convention.
